Quake/pr_ext.c

```diff
@@ -354,6 +354,232 @@ static void PF_anglemod(void)
+		Sys_mkdir(name);
+		file = fopen(name, "w+b");
+		if (file)
+			fseek(file, 0, SEEK_END);
+		break;
+	case 2: //write
+		q_snprintf(name, sizeof(name), "%s/%s", com_gamedir, fname);
+		sl = name + 1;
+		while (*sl)
+		{
+			if (*sl == '/')
+			{
+				*sl = 0;
+				Sys_mkdir(name);	//make sure each part of the path exists.
+				*sl = '/';
+			}
+			sl++;
+		}
+		file = fopen(name, "wb");
+		break;
+	default:
+		Con_Warning("PF_fopen: couldn't open file\n");
+		break;
+	}
+	if (!file)
+		return;
+
+	for (i = 0; ; i++)
+	{
+		if (i == qcfiles_max)
+		{
+			qcfiles_max++;
+			qcfiles = Z_Realloc(qcfiles, sizeof(*qcfiles) * qcfiles_max);
+		}
+		if (!qcfiles[i].file)
+			break;
+	}
+	qcfiles[i].filebase = ftell(file);
+	qcfiles[i].owningvm = qcvm;
+	qcfiles[i].file = file;
+	qcfiles[i].mode = fmode;
+	//reading needs size info
+	qcfiles[i].filesize = filesize;
+	//clear the read cache.
+	qcfiles[i].fileoffset = qcfiles[i].cacheoffset = qcfiles[i].cachesize = 0;
+
+	G_FLOAT(OFS_RETURN) = i + QC_FILE_BASE;
+}
+static void PF_fgets(void)
+{
+	size_t fileid = G_FLOAT(OFS_PARM0) - QC_FILE_BASE;
+	G_INT(OFS_RETURN) = 0;
+	if (fileid >= qcfiles_max)
+		Con_Warning("PF_fgets: invalid file handle\n");
+	else if (!qcfiles[fileid].file)
+		Con_Warning("PF_fgets: file not open\n");
+	else if (qcfiles[fileid].mode != 0)
+		Con_Warning("PF_fgets: file not open for reading\n");
+	else
+	{
+		struct qcfile_s* f = &qcfiles[fileid];
+		char* ret = PR_GetTempString();
+		char* s = ret;
+		char* end = ret + STRINGTEMP_LENGTH;
+		for (;;)
+		{
+			if (f->cacheoffset == f->cachesize)
+			{
+				//figure out how much we can try to cache.
+				int sz = f->filesize - f->fileoffset;
+				if (sz < 0 || f->fileoffset < 0)	//... maybe we shouldn't have implemented seek support.
+					sz = 0;
+				else if ((size_t)sz > sizeof(f->cache))
+					sz = sizeof(f->cache);
+				//read a chunk
+				f->cacheoffset = 0;
+				f->cachesize = fread(f->cache, 1, sz, f->file);
+				f->fileoffset += f->cachesize;
+				if (!f->cachesize)
+				{
+					if (s == ret)
+					{	//absolutely nothing to spew
+						G_INT(OFS_RETURN) = 0;
+						return;
+					}
+					//classic eof...
+					break;
+				}
+			}
+			*s = f->cache[f->cacheoffset++];
+			if (*s == '\n')	//new line, yay!
+				break;
+			s++;
+			if (s == end)
+				s--;	//rewind if we're overflowing, such that we truncate the string.
+		}
+		if (s > ret && s[-1] == '\r')
+			s--;	//terminate it on the \r of a \r\n pair.
+		*s = 0;	//terminate it
+		G_INT(OFS_RETURN) = PR_SetEngineString(ret);
+	}
+}
+static void PF_fclose(void)
+{
+	size_t fileid = G_FLOAT(OFS_PARM0) - QC_FILE_BASE;
+	if (fileid >= qcfiles_max)
+		Con_Warning("PF_fclose: invalid file handle\n");
+	else if (!qcfiles[fileid].file)
+		Con_Warning("PF_fclose: file not open\n");
+	else
+	{
+		fclose(qcfiles[fileid].file);
+		qcfiles[fileid].file = NULL;
+		qcfiles[fileid].owningvm = NULL;
+	}
+}
+static void PF_frikfile_shutdown(void)
```

Comment:
What is this for??

Reply:
I think it was accidentally added, should be removed.

Reply:
It looks like it's supposed to be called in [PR_ShutdownExtensions](https://github.com/Shpoike/Quakespasm/blob/c48936385a4428ce701059a11215780ec3cdcc00/quakespasm/Quake/pr_ext.c#L7706-L7710) to avoid keeping files open across map changes.

Reply:
Yeah, seems like it. Please fix @p4thlight